scripts/benchmark.py

```diff
@@ -0,0 +1,475 @@
+            self.perflogruns.append(self.test.perflog)
+            tot += self.test.dur_secs
+            if (len(self.realruns) >= self._min_runs and tot >= self._cum_dur) or len(self.realruns) >= self._max_runs:
+                break
+
+class BenchComp:
+    def __init__(self, oldbench, newbench, psig=0.05, usertime=False):
+        self.name = oldbench.name
+        self.psig = psig
+        self.old = oldbench.realruns
+        self.new = newbench.realruns
+        if usertime:
+            self.old = oldbench.userruns
+            self.new = newbench.userruns
+
+        self.iqr_old = _iqr(self.old)
+        self.iqr_new = _iqr(self.new)
+        try:
+            result = mannwhitneyu(self.iqr_old, self.iqr_new, alternative='two-sided')
+            self.pvalue = result.pvalue
+        except:
+            self.pvalue = 1.0
+
+        self.u = result[0]
+        self.avg_old = float(sum(self.iqr_old))/len(self.iqr_old)
+        self.avg_new = float(sum(self.iqr_new))/len(self.iqr_new)
+        self.speed_change = (self.avg_new - self.avg_old) / self.avg_old
+
+    @classmethod
+    def header(cls, revold, revnew):
+        oldstr, newstr = revold, revnew
+        if len(oldstr) > 12:
+            oldstr = oldstr[:12]
+        if len(newstr) > 12:
+            newstr = newstr[:12]
+        revstr = ' {} to {} '.format(oldstr, newstr)
+        revstr = revstr.center(30,'-')
+        return '' \
+            + '--------------------------------{}--------------------------------'.format(revstr) \
+            + '\n{:^30s}   {:^15s}   {:^15s}   {:5s}'.format('benchmark', 'old (sec/run)', 'new (sec/run)', 'speedup (pvalue,nsamples)') \
+            + '\n----------------------------------------------------------------------------------------------'
+    @classmethod
+    def footer(cls):
+        return '----------------------------------------------------------------------------------------------'
+
+    def __str__(self):
+        name = self.name
+        if len(name) > 30:
+            name = name[:27] + '...'
+        if self.pvalue <= self.psig:
+            return '{:>30s}:   {:^15f}   {:^15f}   {:+5.1f}% (p={:.3f} n={}+{})'.format(name, self.avg_old, self.avg_new, self.speed_change*100, self.pvalue, len(self.iqr_old), len(self.iqr_new))
+        else:
+            return '{:>30s}:   {:^15f}   {:^15f}      ~   (p={:.3f} n={}+{})'.format(name, self.avg_old, self.avg_new, self.pvalue, len(self.iqr_old), len(self.iqr_new))
+
+def _iqr(a):
+    """return elements of a and be in the interquartile range (inclusive) of both a and b"""
+    qup, qlow = numpy.percentile(a, [75 ,25])
+
+    iqra = []
+    for val in a:
+        if qlow <= val and val <= qup:
+            iqra.append(val)
+    return iqra
+
+
+class BenchmarkDB:
+    def __init__(self, fname):
+        CREATE_BENCH_TABLE = '''CREATE TABLE IF NOT EXISTS benchmarks
+        (
+          id INTEGER PRIMARY KEY AUTOINCREMENT,
+          name TEXT,
+          executable TEXT,
```

Comment:
I'd prefer to throw away the path, I don't think it adds value and could potentially (although unlikely) add issues. However, I do think it's worth splitting off the suffix from this before storing it in the database. Reason: I think it may be useful (or somebody will eventually ask us for it) to compare performance of different methods. We might use this to justify moving from "opt" as the default to "oprof" for workstations for instance, or maybe we can convince developers that "devel" is really where it's at for developing because it's not really slower. Having it as a separate field would make querying the same application for different methods much easier.